File: libs/wasm-sdk/wasm-sdk-nim/src/context.nim
import std/[strutils, tables]
import types, host

type
  Context* = object
    input*: ExecutionInput
    result*: ExecutionResult

proc newContext*(input: ExecutionInput): Context =
  Context(input: input, result: ok())

# -- Input getters --

proc getRaw*(ctx: Context; name: string): string =
  if ctx.input.inputs.hasKey(name):
    ctx.input.inputs[name]
  else:
    ""

proc getString*(ctx: Context; name: string; default: string = ""): string =
  if not ctx.input.inputs.hasKey(name): return default
  let v = ctx.input.inputs[name]
  if v.len >= 2 and v[0] == '"' and v[^1] == '"':
    v[1 ..< v.len - 1]
  else:
    v

proc getI64*(ctx: Context; name: string; default: int64 = 0): int64 =
  if not ctx.input.inputs.hasKey(name): return default
  try:
    parseBiggestInt(ctx.input.inputs[name])
  except ValueError:
    default

proc getF64*(ctx: Context; name: string; default: float64 = 0.0): float64 =
  if not ctx.input.inputs.hasKey(name): return default
  try:
    parseFloat(ctx.input.inputs[name])
  except ValueError:
    default

proc getBool*(ctx: Context; name: string; default: bool = false): bool =
  if not ctx.input.inputs.hasKey(name): return default
  ctx.input.inputs[name] == "true"

# -- Metadata shortcuts --

proc nodeId*(ctx: Context): string = ctx.input.nodeId
proc nodeName*(ctx: Context): string = ctx.input.nodeName
proc runId*(ctx: Context): string = ctx.input.runId
proc appId*(ctx: Context): string = ctx.input.appId
proc boardId*(ctx: Context): string = ctx.input.boardId
proc userId*(ctx: Context): string = ctx.input.userId
proc streamEnabled*(ctx: Context): bool = ctx.input.streamState
proc getLogLevel*(ctx: Context): uint8 = ctx.input.logLevel

# -- Output setters --

proc setOutput*(ctx: var Context; name, jsonValue: string) =
  ctx.result.outputs[name] = jsonValue

proc activateExec*(ctx: var Context; pin: string) =
  ctx.result.activateExec.add pin

proc setPending*(ctx: var Context; p: bool) =
  ctx.result.pending = p

proc setError*(ctx: var Context; msg: string) =
  ctx.result.error = msg

# -- Level-gated logging --

proc trace*(ctx: Context; msg: string) =
  if ctx.input.logLevel <= LogLevelTrace: logTrace(msg)

proc debug*(ctx: Context; msg: string) =
  if ctx.input.logLevel <= LogLevelDebug: logDebug(msg)

proc info*(ctx: Context; msg: string) =
  if ctx.input.logLevel <= LogLevelInfo: logInfo(msg)

proc warn*(ctx: Context; msg: string) =
  if ctx.input.logLevel <= LogLevelWarn: logWarn(msg)

proc error*(ctx: Context; msg: string) =
  if ctx.input.logLevel <= LogLevelError: logError(msg)

# -- Conditional streaming --

proc streamText*(ctx: Context; text: string) =
  if ctx.input.streamState: host.streamText(text)

proc streamJson*(ctx: Context; json: string) =
  if ctx.input.streamState: streamEmit("json", json)

proc streamProgress*(ctx: Context; pct: float; message: string) =
  if ctx.input.streamState:
    let data = "{\"progress\":" & $pct & ",\"message\":" & jsonQuote(message) & "}"
    streamEmit("progress", data)

# -- Variables --

proc getVariable*(ctx: Context; name: string): string =
  varGet(name)

proc setVariable*(ctx: Context; name, value: string) =
  varSet(name, value)

proc deleteVariable*(ctx: Context; name: string) =
  varDelete(name)

proc hasVariable*(ctx: Context; name: string): bool =
  varHas(name)

# -- Storage --

proc readStorage*(ctx: Context; path: string): string =
  storageRead(path)

proc writeStorage*(ctx: Context; path, data: string): int32 =
  storageWrite(path, data)

proc getStorageDir*(ctx: Context; nodeScoped: bool = false): string =
  storageDir(nodeScoped)

proc getUploadDir*(ctx: Context): string =
  uploadDir()

proc getCacheDir*(ctx: Context; nodeScoped: bool = false; userScoped: bool = false): string =
  cacheDir(nodeScoped, userScoped)

proc getUserDir*(ctx: Context; nodeScoped: bool = false): string =
  userDir(nodeScoped)

proc listStorage*(ctx: Context; path: string): string =
  storageList(path)

# -- Models --

proc embedText*(ctx: Context; bit, texts: string): string =
  host.embedText(bit, texts)

# -- HTTP --

proc httpRequest*(ctx: Context; meth: int32; url, headers, body: string): int32 =
  host.httpRequest(meth, url, headers, body)

# -- Finalization --

proc finish*(ctx: var Context): ExecutionResult =
  result = ctx.result

proc success*(ctx: var Context): ExecutionResult =
  ctx.activateExec("exec_out")
  ctx.finish()

proc fail*(ctx: var Context; msg: string): ExecutionResult =
  ctx.setError(msg)
  ctx.finish()

File: libs/wasm-sdk/wasm-sdk-zig/src/context.zig
const std = @import("std");
const types = @import("types.zig");
const host = @import("host.zig");
const mem = @import("memory.zig");

const ExecutionInput = types.ExecutionInput;
const ExecutionResult = types.ExecutionResult;

pub const Context = struct {
    input: ExecutionInput,
    result: ExecutionResult,

    pub fn init(input: ExecutionInput) Context {
        return .{
            .input = input,
            .result = ExecutionResult.success(mem.allocator),
        };
    }

    // --- Metadata ---

    pub fn nodeId(self: *const Context) []const u8 {
        return self.input.node_id;
    }
    pub fn nodeName(self: *const Context) []const u8 {
        return self.input.node_name;
    }
    pub fn runId(self: *const Context) []const u8 {
        return self.input.run_id;
    }
    pub fn appId(self: *const Context) []const u8 {
        return self.input.app_id;
    }
    pub fn boardId(self: *const Context) []const u8 {
        return self.input.board_id;
    }
    pub fn userId(self: *const Context) []const u8 {
        return self.input.user_id;
    }
    pub fn streamEnabled(self: *const Context) bool {
        return self.input.stream_state;
    }
    pub fn logLevelValue(self: *const Context) u8 {
        return self.input.log_level;
    }

    // --- Input getters ---

    pub fn getInput(self: *const Context, name: []const u8) ?[]const u8 {
        return self.input.inputs.get(name);
    }

    pub fn getString(self: *const Context, name: []const u8, default: []const u8) []const u8 {
        const v = self.input.inputs.get(name) orelse return default;
        if (v.len >= 2 and v[0] == '"' and v[v.len - 1] == '"') {
            return v[1 .. v.len - 1];
        }
        return v;
    }

    pub fn getI64(self: *const Context, name: []const u8, default: i64) i64 {
        const v = self.input.inputs.get(name) orelse return default;
        return std.fmt.parseInt(i64, v, 10) catch default;
    }

    pub fn getF64(self: *const Context, name: []const u8, default: f64) f64 {
        const v = self.input.inputs.get(name) orelse return default;
        return std.fmt.parseFloat(f64, v) catch default;
    }

    pub fn getBool(self: *const Context, name: []const u8, default: bool) bool {
        const v = self.input.inputs.get(name) orelse return default;
        return std.mem.eql(u8, v, "true");
    }

    // --- Output setters ---

    pub fn setOutput(self: *Context, name: []const u8, value: []const u8) void {
        self.result.setOutput(name, value);
    }

    pub fn activateExec(self: *Context, pin_name: []const u8) void {
        self.result.activateExecPin(pin_name);
    }

    pub fn setPending(self: *Context, pending: bool) void {
        self.result.pending = pending;
    }

    pub fn setError(self: *Context, err: []const u8) void {
        self.result.error_msg = err;
    }

    // --- Level-gated logging ---

    fn shouldLog(self: *const Context, level: u8) bool {
        return level >= self.input.log_level;
    }

    pub fn debug(self: *const Context, msg: []const u8) void {
        if (self.shouldLog(types.log_level_debug)) host.logDebug(msg);
    }

    pub fn info(self: *const Context, msg: []const u8) void {
        if (self.shouldLog(types.log_level_info)) host.logInfo(msg);
    }

    pub fn warn(self: *const Context, msg: []const u8) void {
        if (self.shouldLog(types.log_level_warn)) host.logWarn(msg);
    }

    pub fn logError(self: *const Context, msg: []const u8) void {
        if (self.shouldLog(types.log_level_error)) host.logError(msg);
    }

    // --- Streaming ---

    pub fn streamText(self: *const Context, txt: []const u8) void {
        if (self.streamEnabled()) host.streamText(txt);
    }

    pub fn streamJson(self: *const Context, data: []const u8) void {
        if (self.streamEnabled()) host.streamEmit("json", data);
    }

    pub fn streamProgress(self: *const Context, progress: f32, message: []const u8) void {
        if (!self.streamEnabled()) return;
        var buf: [256]u8 = undefined;
        const json = std.fmt.bufPrint(&buf, "{{\"progress\":{d},\"message\":\"{s}\"}}", .{ progress, message }) catch return;
        host.streamEmit("progress", json);
    }

    // --- Variables ---

    pub fn getVariable(self: *const Context, name: []const u8) []const u8 {
        _ = self;
        return host.getVariable(name);
    }

    pub fn setVariable(self: *const Context, name: []const u8, value: []const u8) void {
        _ = self;
        host.setVariable(name, value);
    }

    // --- Finalize ---

    pub fn finish(self: *Context) ExecutionResult {
        return self.result;
    }

    pub fn success(self: *Context) ExecutionResult {
        self.activateExec("exec_out");
        return self.finish();
    }

    pub fn fail(self: *Context, err: []const u8) ExecutionResult {
        self.setError(err);
        return self.finish();
    }
};
